kotlin/src/main/java/com/squareup/moshi/KotlinJsonAdapter.kt

```diff
@@ -0,0 +1,141 @@
+@file:Suppress("UNCHECKED_CAST")
+
+package com.squareup.moshi
+
+import com.squareup.moshi.ClassJsonAdapter.includeField
+import com.squareup.moshi.ClassJsonAdapter.isPlatformType
+import java.lang.reflect.Modifier
+import java.lang.reflect.Type
+import kotlin.reflect.KFunction
+import kotlin.reflect.KParameter
+import kotlin.reflect.KProperty1
+import kotlin.reflect.full.declaredMemberProperties
+import kotlin.reflect.full.findAnnotation
+import kotlin.reflect.full.primaryConstructor
+import kotlin.reflect.jvm.isAccessible
+import kotlin.reflect.jvm.javaField
+
+/**
+ * A [JsonAdapter.Factory] for Kotlin data classes. If this is added to Moshi,
+ * data classes will be treated differently from other classes.
+ *
+ * Moshi will call its primary constructor to create the object. This allows for delegated properties
+ * to be initialized correctly. If the JSON does not contain a value for a parameter with a default value
+ * the default value is used. If no default value is present an [IllegalArgumentException] will be
+ * thrown during deserialization.
+ *
+ * Data classes to be (de-)serialized must conform to the following rules:
+ *
+ * - Any properties not found in the primary constructor must be transient, including delegated properties.
+ * - The primary constructor may not contain transient properties unless a default value is provided.
+ * - If the class is a platform type (kotlin.*), it may not contain private fields.
+ */
+class KotlinJsonAdapterFactory : JsonAdapter.Factory {
+  override fun create(type: Type, annotations: Set<Annotation>, moshi: Moshi): JsonAdapter<*>? {
+    val kclass = (type as? Class<*>)?.kotlin ?: return null
+    if (!kclass.isData) return null
+    // This gives us a few guarantees:
+    // 1. Primary constructor exists
+    // 2. Constructor params are var/val and have backing fields
+    // 3. Class is not abstract or an interface
+    // 4. Class is final
+
+    val constructor = kclass.primaryConstructor!!
+    constructor.isAccessible = true
+
+    val concretePropertyByName = kclass.declaredMemberProperties
+        .filter { it.javaField != null }
+        .associateBy { it.name }
+    val constructorParameterByName = constructor.parameters.associateBy { it.name }
+    for ((name, property) in concretePropertyByName) {
+      if (name !in constructorParameterByName.keys &&
+          !Modifier.isTransient(property.javaField!!.modifiers)) {
+        throw IllegalArgumentException("Property $name is not transient. " +
```

Comment:
(this can remain as an IllegalArgumentException. JsonDataException is only for mismatches between data and schema)